File: 61.Rotate-List.py
# https://leetcode.com/problems/rotate-list/
#
# algorithms
# Medium (26.09%)
# Total Accepted:    169,600
# Total Submissions: 650,019

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        length = 0
        tmp = head
        tail = None
        while tmp:
            length += 1
            tail = tmp
            tmp = tmp.next
        
        if length == 0:
            return head
        
        k %= length
        if k == 0:
            return head
        
        tmp = head
        pre = None
        for _ in xrange(length - k):
            pre = tmp
            tmp = tmp.next
        
        res = ListNode(-1)
        res.next = tmp
        pre.next = None
        tail.next = head
        
        return res.next
        

File: 61.Rotate-List.java
// https://leetcode.com/problems/rotate-list/
//
// algorithms
// Medium (27.10%)
// Total Accepted:    191,463
// Total Submissions: 706,379


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode tmp = head, tail = null;
        int len = 0;
        while (tmp != null) {
            len++;
            tail = tmp;
            tmp = tmp.next;
        }
        
        k %= len;
        k = len - k;
        for (int i = 0; i < k; i++) {
            tmp = head.next;
            head.next = null;
            tail.next = head;
            tail = head;
            head = tmp;
        }
        
        return head;
    }
}